### ISSUE TITLE
Running Directed Beam Search with BART

### ISSUE THREAD

**Issue by vnik18** [dapascual/DirectedBeamSearch#4]
Hi, thanks for sharing your code! I have been able to successfully run it with GPT-2 as described in your paper, and I am now trying to run it with other text generation models like BART. 

I have replaced the  `GPT2LMHeadModel` in your code with the `BartForConditionalGeneration` model in the file `main_DBS.py`, and have also replaced the tokenizers. However, I am getting very poor results as the output of the decoder for the 50 sets of 5 keywords dataset that you have provided. The results from BART with Directed Beam Search are much worse than that with GPT-2. In many cases, on using BART the sentence that is formed does not contain all the 5 keywords. 

Do you have any insights on this? Thank you!

**Comment by dapascual** [comment]
Hi, that's interesting. To regulate the appearance of the keywords you need to tune the lambda parameter, you can pass it as an argument when running the code (-weight=<some value>) or directly change it in line 453. The higher that value the more likely the appearance of the words (if it was set to infinity the keywords would appear one after the other). So, since you say that the keywords do not appear with the current value of 20 you probably need to increase that number, so maybe try first with much larger values like 100 or 500 and then tune a bit from there, once you are in the correct order of magnitude the method should not be sensitive to small changes.

I am interested in the results you could get with BART, let me know if this helps and if you calculated some metrics would be great to know how BART+DBS would compare against BART.

I hope this is helpful, let me know if I can help with something else!

**Comment by vnik18** [comment]
Hi, thank you for replying! I tried to increase the value of the weight parameter and unfortunately, that doesn't seem to help either. Most final generated sentences have less than 3 keywords out of 5 in them. 

Another issue I found is that the sentences generated using BART+DBS are not as good in terms of grammar and fluency in English, as compared to GPT-2+DBS. 

Do you have any ideas regarding this? Thank you!

**Comment by dapascual** [comment]
the problem in fluency may be an issue with BART itself (GPT2 is probably a better language generator). How do sentences generated with BART compare to sentences generated with GPT2?

**Comment by vnik18** [comment]
@dapascual No, I checked that, and there is not a significant difference between the sentences generated by BART and those by GPT-2. However, on adding DBS to both models, GPT-2 ends up being significantly better than BART.

**Comment by dapascual** [comment]
Hi, sorry, but I do not have a good answer for this. I think it is an interesting issue but it requires some further investigation of the embedding space of BART in order to understand why modifying its logits affects generation differently from modifying GPT2's logits, which is a good direction for future work. I will keep it in mind.

Thank you for rising this issue.
